File: packages/runtime/src/tools/built-in/calculator.ts
import type { BuiltinTool } from "@llm-space/core";

import type { ToolEntry } from "../tool-registry";

const MAX_EXPRESSION_LENGTH = 1_000;

type Token =
  | { type: "number"; value: number; position: number }
  | { type: "identifier"; value: string; position: number }
  | { type: "operator"; value: string; position: number }
  | { type: "eof"; position: number };

interface MathFunction {
  minArgs: number;
  maxArgs: number;
  evaluate: (...args: number[]) => number;
}

const FUNCTIONS: Record<string, MathFunction> = {
  abs: unary(Math.abs),
  acos: unary(Math.acos),
  asin: unary(Math.asin),
  atan: unary(Math.atan),
  atan2: binary(Math.atan2),
  ceil: unary(Math.ceil),
  cos: unary(Math.cos),
  exp: unary(Math.exp),
  floor: unary(Math.floor),
  log: unary(Math.log),
  log10: unary(Math.log10),
  max: variadic(Math.max),
  min: variadic(Math.min),
  pow: binary(Math.pow),
  round: unary(Math.round),
  sin: unary(Math.sin),
  sqrt: unary(Math.sqrt),
  tan: unary(Math.tan),
  trunc: unary(Math.trunc),
};

const CONSTANTS: Record<string, number> = {
  e: Math.E,
  pi: Math.PI,
};

export const calculatorTool: BuiltinTool = {
  type: "builtin",
  name: "calculator",
  icon: "calculator",
  description:
    "Evaluate a mathematical expression safely. You must use this tool for every task that involves mathematical calculation, even the simplest arithmetic; never calculate the result yourself. Supports parentheses, +, -, *, /, %, **, constants pi/e, and common functions such as sin, cos, tan, sqrt, log, min, and max. Math.* names are also accepted.",
  strict: true,
  parameters: {
    type: "object",
    required: ["expression"],
    properties: {
      expression: {
        type: "string",
        description:
          "The mathematical expression to evaluate, for example: 2 * (3 + 4), Math.sin(Math.PI / 2), or sqrt(81).",
      },
    },
    additionalProperties: false,
  },
};

export function calculate(expression: string): number {
  if (typeof expression !== "string") {
    throw new Error("expression must be a string.");
  }
  if (!expression.trim()) {
    throw new Error("expression must not be empty.");
  }
  if (expression.length > MAX_EXPRESSION_LENGTH) {
    throw new Error(
      `expression exceeds the ${MAX_EXPRESSION_LENGTH} character limit.`
    );
  }

  try {
    return new Parser(tokenize(expression)).parse();
  } catch (error) {
    if (error instanceof CalculatorError) {
      throw new Error(`Calculator ${error.kind} error: ${error.message}`, {
        cause: error,
      });
    }
    throw error;
  }
}

export const calculatorBuiltInTools: ToolEntry[] = [
  {
    tool: calculatorTool,
    execute(args: Record<string, unknown>) {
      if (typeof args.expression !== "string") {
        return Promise.reject(new Error("expression must be a string."));
      }
      return Promise.resolve(calculate(args.expression));
    },
  },
];

class CalculatorError extends Error {
  constructor(
    readonly kind: "syntax" | "range",
    message: string
  ) {
    super(message);
  }
}

class Parser {
  private index = 0;

  constructor(private readonly tokens: Token[]) {}

  parse(): number {
    const value = this.parseAdditive();
    const token = this.peek();
    if (token.type !== "eof") {
      this.syntax(`unexpected token ${describeToken(token)}`, token.position);
    }
    return finite(value);
  }

  private parseAdditive(): number {
    let value = this.parseMultiplicative();
    while (this.matches("+") || this.matches("-")) {
      const operator = this.previous().value;
      const right = this.parseMultiplicative();
      value = finite(operator === "+" ? value + right : value - right);
    }
    return value;
  }

  private parseMultiplicative(): number {
    let value = this.parseUnary();
    while (
      this.matches("*") ||
      this.matches("/") ||
      this.matches("%")
    ) {
      const operator = this.previous().value;
      const right = this.parseUnary();
      if ((operator === "/" || operator === "%") && right === 0) {
        throw new CalculatorError("range", "division by zero.");
      }
      value = finite(
        operator === "*"
          ? value * right
          : operator === "/"
            ? value / right
            : value % right
      );
    }
    return value;
  }

  private parseUnary(): number {
    if (this.matches("+")) {
      return this.parseUnary();
    }
    if (this.matches("-")) {
      return finite(-this.parseUnary());
    }
    return this.parsePower();
  }

  private parsePower(): number {
    let value = this.parsePrimary();
    if (this.matches("**")) {
      value = finite(Math.pow(value, this.parseUnary()));
    }
    return value;
  }

  private parsePrimary(): number {
    const token = this.advance();
    if (token.type === "number") {
      return token.value;
    }
    if (token.type === "operator" && token.value === "(") {
      const value = this.parseAdditive();
      this.consume(")", "expected ')' to close the expression");
      return value;
    }
    if (token.type === "identifier") {
      return this.parseIdentifier(token);
    }
    this.syntax(`expected a number, function, or '('`, token.position);
  }

  private parseIdentifier(token: Extract<Token, { type: "identifier" }>): number {
    const name = normalizeIdentifier(token.value);
    if (!this.matches("(")) {
      const constant = CONSTANTS[name];
      if (constant === undefined) {
        this.syntax(`unknown constant '${token.value}'`, token.position);
      }
      return constant;
    }

    const fn = FUNCTIONS[name];
    if (!fn) {
      this.syntax(`unknown function '${token.value}'`, token.position);
    }
    const args: number[] = [];
    if (!this.check(")")) {
      do {
        args.push(this.parseAdditive());
      } while (this.matches(","));
    }
    this.consume(")", `expected ')' after arguments to '${token.value}'`);
    if (args.length < fn.minArgs || args.length > fn.maxArgs) {
      const expected =
        fn.minArgs === fn.maxArgs
          ? `${fn.minArgs}`
          : `${fn.minArgs} or more`;
      this.syntax(
        `'${token.value}' expects ${expected} argument${fn.minArgs === 1 && fn.maxArgs === 1 ? "" : "s"}, got ${args.length}`,
        token.position
      );
    }
    return finite(fn.evaluate(...args));
  }

  private matches(value: string): boolean {
    if (!this.check(value)) return false;
    this.index += 1;
    return true;
  }

  private check(value: string): boolean {
    const token = this.peek();
    return token.type === "operator" && token.value === value;
  }

  private consume(value: string, message: string): void {
    if (!this.matches(value)) this.syntax(message, this.peek().position);
  }

  private advance(): Token {
    const token = this.peek();
    if (token.type !== "eof") this.index += 1;
    return token;
  }

  private previous(): Extract<Token, { type: "operator" }> {
    return this.tokens[this.index - 1] as Extract<Token, { type: "operator" }>;
  }

  private peek(): Token {
    return this.tokens[this.index];
  }

  private syntax(message: string, position: number): never {
    throw new CalculatorError("syntax", `${message} at position ${position}.`);
  }
}

function tokenize(expression: string): Token[] {
  const tokens: Token[] = [];
  let index = 0;
  while (index < expression.length) {
    const char = expression[index];
    if (/\s/.test(char)) {
      index += 1;
      continue;
    }
    if (/\d/.test(char) || (char === "." && /\d/.test(expression[index + 1] ?? ""))) {
      const match = /^(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?/.exec(
        expression.slice(index)
      );
      if (!match) {
        throw new CalculatorError(
          "syntax",
          `invalid number at position ${index}.`
        );
      }
      const value = Number(match[0]);
      if (!Number.isFinite(value)) {
        throw new CalculatorError("range", `number is outside the supported range at position ${index}.`);
      }
      tokens.push({ type: "number", value, position: index });
      index += match[0].length;
      continue;
    }
    if (/[A-Za-z_]/.test(char)) {
      const match = /^[A-Za-z_][A-Za-z0-9_.]*/.exec(
        expression.slice(index)
      );
      if (!match) {
        throw new CalculatorError(
          "syntax",
          `invalid identifier at position ${index}.`
        );
      }
      tokens.push({ type: "identifier", value: match[0], position: index });
      index += match[0].length;
      continue;
    }
    const twoCharacters = expression.slice(index, index + 2);
    if (twoCharacters === "**") {
      tokens.push({ type: "operator", value: twoCharacters, position: index });
      index += 2;
      continue;
    }
    if ("+-*/%(),".includes(char)) {
      tokens.push({ type: "operator", value: char, position: index });
      index += 1;
      continue;
    }
    throw new CalculatorError(
      "syntax",
      `unsupported character '${char}' at position ${index}.`
    );
  }
  tokens.push({ type: "eof", position: expression.length });
  return tokens;
}

function normalizeIdentifier(identifier: string): string {
  return identifier.startsWith("Math.")
    ? identifier.slice("Math.".length).toLowerCase()
    : identifier.toLowerCase();
}

function finite(value: number): number {
  if (!Number.isFinite(value)) {
    throw new CalculatorError(
      "range",
      "result is not a finite number (overflow or invalid function domain)."
    );
  }
  return value;
}

function describeToken(token: Token): string {
  if (token.type === "eof") return "end of expression";
  return `'${token.value}'`;
}

function unary(evaluate: (value: number) => number): MathFunction {
  return { minArgs: 1, maxArgs: 1, evaluate };
}

function binary(evaluate: (left: number, right: number) => number): MathFunction {
  return { minArgs: 2, maxArgs: 2, evaluate };
}

function variadic(evaluate: (...values: number[]) => number): MathFunction {
  return { minArgs: 1, maxArgs: Number.POSITIVE_INFINITY, evaluate };
}

File: packages/core/src/generator/langgraph/tools/built-in/calculator.py
import math
import re
from dataclasses import dataclass
from typing import Callable

from langchain.tools import tool


MAX_EXPRESSION_LENGTH = 1_000


class CalculatorError(ValueError):
    def __init__(self, kind: str, message: str):
        super().__init__(message)
        self.kind = kind


@dataclass(frozen=True)
class Token:
    type: str
    value: object
    position: int


@dataclass(frozen=True)
class MathFunction:
    min_args: int
    max_args: float
    evaluate: Callable[..., float]


def _unary(fn: Callable[[float], float]) -> MathFunction:
    return MathFunction(1, 1, fn)


def _binary(fn: Callable[[float, float], float]) -> MathFunction:
    return MathFunction(2, 2, fn)


def _variadic(fn: Callable[..., float]) -> MathFunction:
    return MathFunction(1, math.inf, fn)


def _js_round(value: float) -> float:
    return math.floor(value + 0.5)


FUNCTIONS = {
    "abs": _unary(abs),
    "acos": _unary(math.acos),
    "asin": _unary(math.asin),
    "atan": _unary(math.atan),
    "atan2": _binary(math.atan2),
    "ceil": _unary(math.ceil),
    "cos": _unary(math.cos),
    "exp": _unary(math.exp),
    "floor": _unary(math.floor),
    "log": _unary(math.log),
    "log10": _unary(math.log10),
    "max": _variadic(max),
    "min": _variadic(min),
    "pow": _binary(math.pow),
    "round": _unary(_js_round),
    "sin": _unary(math.sin),
    "sqrt": _unary(math.sqrt),
    "tan": _unary(math.tan),
    "trunc": _unary(math.trunc),
}

CONSTANTS = {"e": math.e, "pi": math.pi}
NUMBER_RE = re.compile(r"^(?:\d+(?:\.\d*)?|\.\d+)(?:[eE][+-]?\d+)?")
IDENTIFIER_RE = re.compile(r"^[A-Za-z_][A-Za-z0-9_.]*")


def _finite(value: float) -> float:
    if not math.isfinite(value):
        raise CalculatorError(
            "range",
            "result is not a finite number (overflow or invalid function domain).",
        )
    return value


def _tokenize(expression: str) -> list[Token]:
    tokens = []
    index = 0
    while index < len(expression):
        char = expression[index]
        if char.isspace():
            index += 1
            continue
        if char.isdigit() or (
            char == "."
            and index + 1 < len(expression)
            and expression[index + 1].isdigit()
        ):
            match = NUMBER_RE.match(expression[index:])
            if not match:
                raise CalculatorError("syntax", f"invalid number at position {index}.")
            text = match.group(0)
            value = float(text)
            if not math.isfinite(value):
                raise CalculatorError(
                    "range",
                    f"number is outside the supported range at position {index}.",
                )
            tokens.append(Token("number", value, index))
            index += len(text)
            continue
        if char.isascii() and (char.isalpha() or char == "_"):
            match = IDENTIFIER_RE.match(expression[index:])
            assert match is not None
            value = match.group(0)
            tokens.append(Token("identifier", value, index))
            index += len(value)
            continue
        if expression[index : index + 2] == "**":
            tokens.append(Token("operator", "**", index))
            index += 2
            continue
        if char in "+-*/%(),":
            tokens.append(Token("operator", char, index))
            index += 1
            continue
        raise CalculatorError(
            "syntax", f"unsupported character '{char}' at position {index}."
        )
    tokens.append(Token("eof", None, len(expression)))
    return tokens


class Parser:
    def __init__(self, tokens: list[Token]):
        self.tokens = tokens
        self.index = 0

    def parse(self) -> float:
        value = self.parse_additive()
        token = self.peek()
        if token.type != "eof":
            self.syntax(f"unexpected token {self.describe(token)}", token.position)
        return _finite(value)

    def parse_additive(self) -> float:
        value = self.parse_multiplicative()
        while self.matches("+") or self.matches("-"):
            operator = self.previous().value
            right = self.parse_multiplicative()
            value = _finite(value + right if operator == "+" else value - right)
        return value

    def parse_multiplicative(self) -> float:
        value = self.parse_unary()
        while self.matches("*") or self.matches("/") or self.matches("%"):
            operator = self.previous().value
            right = self.parse_unary()
            if operator in ("/", "%") and right == 0:
                raise CalculatorError("range", "division by zero.")
            if operator == "*":
                value = _finite(value * right)
            elif operator == "/":
                value = _finite(value / right)
            else:
                value = _finite(math.fmod(value, right))
        return value

    def parse_unary(self) -> float:
        if self.matches("+"):
            return self.parse_unary()
        if self.matches("-"):
            return _finite(-self.parse_unary())
        return self.parse_power()

    def parse_power(self) -> float:
        value = self.parse_primary()
        if self.matches("**"):
            try:
                value = _finite(math.pow(value, self.parse_unary()))
            except (OverflowError, ValueError):
                raise CalculatorError(
                    "range",
                    "result is not a finite number (overflow or invalid function domain).",
                ) from None
        return value

    def parse_primary(self) -> float:
        token = self.advance()
        if token.type == "number":
            return token.value
        if token.type == "operator" and token.value == "(":
            value = self.parse_additive()
            self.consume(")", "expected ')' to close the expression")
            return value
        if token.type == "identifier":
            return self.parse_identifier(token)
        self.syntax("expected a number, function, or '('", token.position)

    def parse_identifier(self, token: Token) -> float:
        name = str(token.value)
        normalized = name[5:].lower() if name.startswith("Math.") else name.lower()
        if not self.matches("("):
            if normalized not in CONSTANTS:
                self.syntax(f"unknown constant '{name}'", token.position)
            return CONSTANTS[normalized]

        fn = FUNCTIONS.get(normalized)
        if fn is None:
            self.syntax(f"unknown function '{name}'", token.position)
        args = []
        if not self.check(")"):
            while True:
                args.append(self.parse_additive())
                if not self.matches(","):
                    break
        self.consume(")", f"expected ')' after arguments to '{name}'")
        if len(args) < fn.min_args or len(args) > fn.max_args:
            expected = (
                str(fn.min_args)
                if fn.min_args == fn.max_args
                else f"{fn.min_args} or more"
            )
            suffix = "" if fn.min_args == 1 and fn.max_args == 1 else "s"
            self.syntax(
                f"'{name}' expects {expected} argument{suffix}, got {len(args)}",
                token.position,
            )
        try:
            return _finite(float(fn.evaluate(*args)))
        except (OverflowError, ValueError):
            raise CalculatorError(
                "range",
                "result is not a finite number (overflow or invalid function domain).",
            ) from None

    def matches(self, value: str) -> bool:
        if not self.check(value):
            return False
        self.index += 1
        return True

    def check(self, value: str) -> bool:
        token = self.peek()
        return token.type == "operator" and token.value == value

    def consume(self, value: str, message: str) -> None:
        if not self.matches(value):
            self.syntax(message, self.peek().position)

    def advance(self) -> Token:
        token = self.peek()
        if token.type != "eof":
            self.index += 1
        return token

    def previous(self) -> Token:
        return self.tokens[self.index - 1]

    def peek(self) -> Token:
        return self.tokens[self.index]

    @staticmethod
    def describe(token: Token) -> str:
        return "end of expression" if token.type == "eof" else f"'{token.value}'"

    @staticmethod
    def syntax(message: str, position: int):
        raise CalculatorError("syntax", f"{message} at position {position}.")


@tool
def calculator(expression: str) -> float:
    """Evaluate a mathematical expression safely.

    You must use this tool for every task that involves mathematical
    calculation, even the simplest arithmetic; never calculate the result
    yourself.

    Supports parentheses, +, -, *, /, %, **, constants pi/e, and common
    functions such as sin, cos, tan, sqrt, log, min, and max. Math.* names are
    also accepted.

    Args:
        expression: The expression to evaluate, for example ``2 * (3 + 4)``,
            ``Math.sin(Math.PI / 2)``, or ``sqrt(81)``.
    """
    if not isinstance(expression, str):
        raise ValueError("expression must be a string.")
    if not expression.strip():
        raise ValueError("expression must not be empty.")
    if len(expression) > MAX_EXPRESSION_LENGTH:
        raise ValueError(
            f"expression exceeds the {MAX_EXPRESSION_LENGTH} character limit."
        )
    try:
        return Parser(_tokenize(expression)).parse()
    except CalculatorError as error:
        raise ValueError(f"Calculator {error.kind} error: {error}") from None
